File: 10.searchInRotatedArray.ts
// rotated array:- [3, 4, 5, 6, 0, 1, 2]
import { question } from "readline-sync";

export function getIndexOfTargetFromBS(
  arr: number[],
  target: number,
  start: number,
  end: number
): number {
  let mid: number;
  while (start <= end) {
    mid = Math.floor((start + end) / 2);

    const midElement = arr[mid];
    if (midElement === target) {
      return mid;
    } else if (midElement > target) {
      end = mid - 1;
    } else {
      start = mid + 1;
    }
  }
  return -1;
}

export function findPivotElementInRotatedArray(arr: number[]): number {
  let start = 0,
    end = arr.length - 1,
    mid: number;

  while (start <= end) {
    mid = Math.floor((start + end) / 2);

    if (arr[mid] > arr[mid + 1]) {
      return mid;
    } else if (arr[start] > arr[mid]) {
      end = mid - 1;
    } else {
      start = mid + 1;
    }
  }
  return -1;
}

export function searchInRotatedArray(arr: number[], target: number): number {
  const pivot = findPivotElementInRotatedArray(arr);
  console.log({ pivot });

  const firstPart = { start: 0, end: pivot };
  const indexInFirstPart = getIndexOfTargetFromBS(
    arr,
    target,
    firstPart.start,
    firstPart.end
  );
  if (indexInFirstPart !== -1) {
    return indexInFirstPart;
  }

  const secondPart = { start: pivot + 1, end: arr.length - 1 };
  const indexInSecondPart = getIndexOfTargetFromBS(
    arr,
    target,
    secondPart.start,
    secondPart.end
  );
  return indexInSecondPart;
}

const rotatedArr = question("Enter rotated array:\n")
  .split(" ")
  .map((s: string) => parseInt(s));
const target = parseInt(question("Enter target: "));
const targetIndex = searchInRotatedArray(rotatedArr, target);
console.log({ targetIndex });

File: 1.Binary Search/10.searchInRotatedArray.py
# https://leetcode.com/problems/search-in-rotated-sorted-array


class Solution:
    def search(self, nums: list[int], target: int) -> int:
        n = len(nums)

        def find_pivot():
            low, high = 0, n - 1
            while low < high:
                mid = (low + high) >> 1
                if nums[mid] > nums[high]:
                    low = mid + 1
                else:
                    high = mid
            return low

        def bs(low: int, high: int) -> int:
            while low <= high:
                mid = (low + high) >> 1
                if nums[mid] == target:
                    return mid
                if nums[mid] < target:
                    low = mid + 1
                else:
                    high = mid - 1
            return -1

        pivot = find_pivot()
        first_part = bs(0, pivot - 1)
        if first_part != -1:
            return first_part
        second_part = bs(pivot, n - 1)
        return second_part


arr = list(map(int, input("Enter arr: ").split()))
k = int(input("Enter target: "))
print(Solution().search(arr, k))
